src/sources/kubernetes/message_parser.rs

```diff
@@ -0,0 +1,144 @@
+/// despite CRI defining it's own format.
+pub fn build_message_parser() -> crate::Result<ApplicableTransform> {
+    let transforms = vec![
+        Box::new(DockerMessageTransformer::new()) as Box<dyn Transform>,
+        transform_cri_message()?,
+    ];
+    Ok(ApplicableTransform::Candidates(transforms))
+}
+
+#[derive(Debug)]
+struct DockerMessageTransformer {
+    json_parser: JsonParser,
+    atom_time: Atom,
+    atom_log: Atom,
+}
+
+impl DockerMessageTransformer {
+    fn new() -> Self {
+        let mut config = JsonParserConfig::default();
+
+        // Drop so that it's possible to detect if message is in json format
+        config.drop_invalid = true;
+
+        config.drop_field = true;
+
+        DockerMessageTransformer {
+            json_parser: config.into(),
+            atom_time: Atom::from("time"),
+            atom_log: Atom::from("log"),
+        }
+    }
+}
+
+impl Transform for DockerMessageTransformer {
+    fn transform(&mut self, event: Event) -> Option<Event> {
+        let mut event = self.json_parser.transform(event)?;
+
+        // Rename fields
+        let log = event.as_mut_log();
+
+        // time -> timestamp
+        if let Some(ValueKind::Bytes(timestamp_bytes)) = log.remove(&self.atom_time) {
+            match DateTime::parse_from_rfc3339(
+                String::from_utf8_lossy(timestamp_bytes.as_ref()).as_ref(),
+            ) {
+                Ok(timestamp) => {
+                    log.insert_explicit(event::TIMESTAMP.clone(), timestamp.with_timezone(&Utc))
+                }
+                Err(error) => {
+                    debug!(message = "Non rfc3339 timestamp.", %error, rate_limit_secs = 10);
+                    return None;
+                }
+            }
+        } else {
+            debug!(message = "Missing field.", field = %self.atom_time, rate_limit_secs = 10);
+            return None;
+        }
+
+        // log -> message
+        if let Some(message) = log.remove(&self.atom_log) {
+            log.insert_explicit(event::MESSAGE.clone(), message);
+        } else {
+            debug!(message = "Missing field.", field = %self.atom_log, rate_limit_secs = 10);
+            return None;
+        }
+
+        Some(event)
+    }
+}
+
+/// As defined by CRI
+fn transform_cri_message() -> crate::Result<Box<dyn Transform>> {
+    let mut rp_config = RegexParserConfig::default();
+    // message field
+    rp_config.regex =
+        r"^(?P<timestamp>.*) (?P<stream>(stdout|stderr)) (?P<multiline_tag>(P|F)) (?P<message>.*)$"
+            .to_owned();
+    // drop field
+    rp_config
+        .types
+        .insert(event::TIMESTAMP.clone(), "timestamp|%+".to_owned());
+    // stream is a string
+    // message is a string
+    RegexParser::build(&rp_config).map_err(|e| {
+        format!(
+            "Failed in creating message regex transform with error: {:?}",
+            e
+        )
+        .into()
+    })
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
we should probably add more tests here related to the docker transform?

Reply:
It's not related to this PR, but I agree.